pandas/core/categorical.py

```diff
@@ -248,22 +249,30 @@ class Categorical(PandasObject):
     __array_priority__ = 1000
     _typ = 'categorical'
 
-    def __init__(self, values, categories=None, ordered=False, fastpath=False):
+    def __init__(self, values, categories=None, ordered=None, dtype=None,
+                 fastpath=False):
 
-        self._validate_ordered(ordered)
+        if dtype is not None:
+            if categories is not None or ordered is not None:
+                raise ValueError("Cannot specify both `dtype` and `categories`"
+                                 " or `ordered`.")
+            categories = dtype.categories
+            ordered = dtype.ordered
+
+        if ordered is None:
+            ordered = False
 
         if fastpath:
-            # fast path
+            if dtype is None:
+                dtype = CategoricalDtype(categories, ordered)
             self._codes = coerce_indexer_dtype(values, categories)
```

Comment:
IOW, you can move this *above* the dtype since fastpath just generates this

Reply:
Moving this up made the code further down messier, since the `categories` passed to `CategoricalDtype` need to be inferred from the `values`.